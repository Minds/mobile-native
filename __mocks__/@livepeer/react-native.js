// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
