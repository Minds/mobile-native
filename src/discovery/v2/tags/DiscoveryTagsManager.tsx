import { observer, useLocalStore } from 'mobx-react';
import React, { useEffect, useCallback } from 'react';
import {
  View,
  StyleProp,
  ViewStyle,
  SectionList,
  SectionListData,
  Platform,
} from 'react-native';
import { useDiscoveryV2Store } from '../DiscoveryV2Context';
import ThemedStyles from '../../../styles/ThemedStyles';

import BottomOptionPopup from '../../../common/components/BottomOptionPopup';
import MenuItem from '../../../common/components/menus/MenuItem';
import MenuSubtitle from '../../../common/components/menus/MenuSubtitle';
import { TDiscoveryTagsTag } from '../DiscoveryV2Store';
import { ScrollView, TouchableOpacity } from 'react-native-gesture-handler';
import InputContainer from '../../../common/components/InputContainer';
import i18n from '../../../common/services/i18n.service';
import { useKeyboard } from '@react-native-community/hooks';

const isIos = Platform.OS === 'ios';

interface Props {
  style?: StyleProp<ViewStyle>;
  show: boolean;
  onCancel: () => void;
  onDone: () => void;
}

/**
 * Key extractor
 */
const keyExtractor = (item) => String(item.value);

const createStore = () => ({
  other: [] as TDiscoveryTagsTag[],
  selected: [] as TDiscoveryTagsTag[],
  deselected: [] as TDiscoveryTagsTag[],
  inputValue: '',
  setValue(v: string) {
    this.inputValue = v;
  },
  createTag() {
    const selectedIndex = this.selected.findIndex(
      (t) => t.value === this.inputValue,
    );
    if (selectedIndex < 0) {
      this.selected.push({ value: this.inputValue });
    }
    this.inputValue = '';
  },
  setTags(selected, other) {
    this.selected = selected;
    this.other = other;
  },
  selectTag(tag) {
    // Queue for adding
    const selectedIndex = this.selected.findIndex((t) => t.value === tag.value);
    if (selectedIndex < 0) {
      this.selected.push(tag);
    }

    // Remove from deselected
    const deselectedIndex = this.deselected.findIndex(
      (t) => t.value === tag.value,
    );
    this.deselected.splice(deselectedIndex, 1);

    // Remove from other
    const otherIndex = this.other.findIndex((t) => t.value === tag.value);
    this.other.splice(otherIndex, 1);
  },
  deselectTag(tag) {
    // Queue for deleting
    const deselectedIndex = this.deselected.findIndex(
      (t) => t.value === tag.value,
    );
    if (deselectedIndex < 0) {
      this.deselected.push(tag);
    }

    // Remove from selected
    const selectedIndex = this.selected.findIndex((t) => t.value === tag.value);
    this.selected.splice(selectedIndex, 1);
  },
});

type StoreType = ReturnType<typeof createStore>;

/**
 * Discovery Manage Tags
 */
export const DiscoveryTagsManager = observer((props: Props) => {
  const theme = ThemedStyles.style;
  const discoveryV2 = useDiscoveryV2Store();
  const store = useLocalStore<StoreType>(createStore);
  const keyboard = useKeyboard();

  useEffect(() => {
    store.setTags(discoveryV2.tags.slice(), discoveryV2.trendingTags.slice());
  }, [discoveryV2.tags, discoveryV2.trendingTags, props.show, store]);

  const ItemPartial = ({ item }) => {
    const tag = item;

    const selected =
      store.selected.findIndex((t) => t.value === tag.value) > -1;

    return (
      <MenuItem
        component={Platform.OS === 'android' ? TouchableOpacity : undefined}
        item={{
          title: '#' + tag.value,
          onPress: () => {
            if (selected) {
              store.deselectTag(tag);
            } else {
              store.selectTag(tag);
            }
          },
          icon: selected
            ? {
                name: 'ios-remove-circle-outline',
                type: 'ionicon',
              }
            : { name: 'ios-add-circle-outline', type: 'ionicon' },
        }}
      />
    );
  };

  const SectionHeaderPatrial = (info: { section: SectionListData<any> }) => {
    return (
      <View style={[theme.backgroundSecondary]}>
        <MenuSubtitle>{info.section.title.toUpperCase()}</MenuSubtitle>
      </View>
    );
  };

  const onCancel = useCallback(() => {
    props.onCancel();
  }, [props]);

  const onDone = useCallback(() => {
    discoveryV2.saveTags(store.selected, store.deselected);
    props.onDone();
  }, [discoveryV2, store, props]);

  const onCreate = useCallback(() => {
    store.createTag();
    discoveryV2.saveTags(store.selected, store.deselected);
  }, [store, discoveryV2]);

  /**
   * Render
   */

  return (
    <BottomOptionPopup
      height={500}
      title={i18n.t('discovery.manage')}
      show={props.show}
      onCancel={onCancel}
      onDone={onDone}
      contentContainerStyle={
        keyboard.keyboardShown && !isIos
          ? { paddingTop: keyboard.keyboardHeight / 2 }
          : undefined
      }
      content={
        <ScrollView>
          <SectionList
            ListHeaderComponent={
              <InputContainer
                placeholder={i18n.t('add')}
                onChangeText={store.setValue}
                onSubmitEditing={onCreate}
                value={store.inputValue}
                style={[
                  theme.backgroundPrimary,
                  theme.marginRight4x,
                  theme.paddingLeft2x,
                ]}
                testID="hashtagInput"
                selectTextOnFocus={true}
              />
            }
            renderItem={ItemPartial}
            renderSectionHeader={SectionHeaderPatrial}
            sections={[
              {
                title: i18n.t('discovery.yourTags'),
                data: store.selected.slice(),
              },
              {
                title: i18n.t('discovery.otherTags'),
                data: store.other.slice(),
              },
            ]}
            keyExtractor={keyExtractor}
            stickySectionHeadersEnabled={true}
          />
        </ScrollView>
      }
      doneText={i18n.t('done')}
    />
  );
});

export default DiscoveryTagsManager;
