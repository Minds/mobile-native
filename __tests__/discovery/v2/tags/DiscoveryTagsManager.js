import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import DiscoveryTagsManager from '../../../../src/discovery/v2/tags/DiscoveryTagsManager';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

/**
 * Tests
 */
describe('DiscoveryTagsManager component', () => {
  it('should render correctly', () => {
    const component = shallow(
      <DiscoveryTagsManager
        show={false}
        onClose={() => {}}
        onDone={() => {}}
      />,
    );

    expect(component).toMatchSnapshot();
  });
});
