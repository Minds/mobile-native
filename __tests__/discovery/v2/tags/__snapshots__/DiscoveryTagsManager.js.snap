// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscoveryTagsManager component should render correctly 1`] = `
<Memo(wrappedComponent)
  content={
    <ScrollView>
      <SectionList
        ListHeaderComponent={
          <InputContainer
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Add"
            selectTextOnFocus={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
              ]
            }
            testID="hashtagInput"
            value=""
          />
        }
        data={Array []}
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onEndReachedThreshold={2}
        renderItem={[Function]}
        renderSectionHeader={[Function]}
        scrollEventThrottle={50}
        sections={
          Array [
            Object {
              "data": Array [],
              "title": "Your tags",
            },
            Object {
              "data": Array [],
              "title": "Other tags",
            },
          ]
        }
        stickySectionHeadersEnabled={true}
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />
    </ScrollView>
  }
  doneText="Done"
  height={500}
  onCancel={[Function]}
  onDone={[Function]}
  show={false}
  title="Manage Tags"
/>
`;
